#### Project Type: Exploratory Data Analysis
### Project Number: 1
## Project Title: Разведочный анализ результатов измерений показания качества электрической энергии ТП Сохондо - Фидер контактной сети №4 

---


# **STAGE 1. PROBLEM DEFINITION**
    




### Project Objectives

---

- Исследовать характер распределения и взаимоотношений между показателями оцениваемыми резервной ступенью дистанционной защиты фидеров контакной сетеи 
- Определить возможные причины неселективной работы третьей ступени направленной дистанционной защиты фидеров конатакной сети 

### Success criteria

---

Выявляение закономерностей позволяющих идентифицировать режим работы СТЭ и однозначно разграничить аварийный и нормальный режим работы. 

### Termenology

---

- ФКС - фидер контактной сети 
- НДЗ - направленная дистанционна защита 
- СТЭ - система тягвого электроснабжения 

### Task type

---

По своей сути перед исследованием стоит решение задачи бинарной классификации: нормальный или аварийны режим работы СТЭ

    
# **STAGE 2. PRELIMINARY DATA WRANGLING**



### Import Libraries

---

In [3]:
# Data manipulation
import pandas as pd
import numpy as np

# Data Visualizations
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

# Time series 
import datetime

### Import Data

---

In [4]:
data_path = 'D:/Результаты измерений СТЭ - Сентябрь 2020/0758 - ТП Сохондо - ФКС 4 (ЦЗА)/0758_ТП_Сохондо_ФКС_4_ЦЗА_EDA.csv'

# data_path = '/media/ruslan/Data/Результаты измерений СТЭ - Сентябрь 2020/0758 - ТП Сохондо - ФКС 4 (ЦЗА)/0758_ТП_Сохондо_ФКС_4_ЦЗА_EDA.csv'

## SECTION 2.1.  DATA UNDERSTANDING

### _Collect initial data_

---

In [5]:
# Загрузим набор даннх для первичного визуального анализа 
fks_4 = pd.read_csv(data_path, sep=';', skiprows=1)
fks_4.head()

дата        Время (UTC+0)  U L1 мгн [kВ]  I L1 мгн [A]  CF U L1 мгн [---]  \
0  08.09.2020  04:22:50.242           14.53        211.92               1.59   
1  08.09.2020  04:22:51.243           14.51        216.06               1.59   
2  08.09.2020  04:22:52.243           14.50        217.02               1.59   
3  08.09.2020  04:22:53.244           14.50        215.73               1.59   
4  08.09.2020  04:22:54.244           14.49        213.97               1.60   

   CF I L1 мгн [---]  THD U L1 мгн [%]  THD I L1 мгн [%]  U H 1 L1 мгн [%]  \
0               1.56              8.86             16.19             99.60   
1               1.58              9.04             15.53             99.58   
2               1.56              9.13             15.46             99.57   
3               1.55              9.11             15.56             99.57   
4               1.57              9.10             15.94             99.58   

   U H 2 L1 мгн [%]  ...  I H 32 L1 мгн [%]  I H 33 L1 мгн [%]  \
0              0.06  ...               0.01               0.24   
1              0.07  ...               0.02               0.30   
2              0.05  ...               0.04               0.31   
3              0.07  ...               0.02               0.37   
4              0.05  ...               0.03               0.37   

   I H 34 L1 мгн [%]  I H 35 L1 мгн [%]  I H 36 L1 мгн [%]  I H 37 L1 мгн [%]  \
0               0.03               0.11               0.03               0.20   
1               0.03               0.17               0.03               0.22   
2               0.03               0.25               0.03               0.19   
3               0.04               0.34               0.03               0.30   
4               0.02               0.12               0.03               0.14   

   I H 38 L1 мгн [%]  I H 39 L1 мгн [%]  I H 40 L1 мгн [%]  \
0               0.02               0.23               0.02   
1               0.03               0.26               0.02   
2               0.02               0.27               0.05   
3               0.02               0.29               0.03   
4               0.04               0.22               0.02   

   cos(φ) L1 мгн [---]  
0                 0.50  
1                 0.50  
2                 0.50  
3                 0.50  
4                 0.49  

[5 rows x 89 columns]

In [6]:
# Форма исходного набора данных
print("Форма исходного набора данных:\n{}".format(fks_4.shape))

Форма исходного набора данных:
(999610, 89)


In [7]:
# Список переменных
fks_4.columns.tolist()

['дата      ',
 'Время (UTC+0)',
 'U L1 мгн [kВ]',
 'I L1 мгн [A]',
 'CF U L1 мгн [---]',
 'CF I L1 мгн [---]',
 'THD U L1 мгн [%]',
 'THD I L1 мгн [%]',
 'U H 1 L1 мгн [%]',
 'U H 2 L1 мгн [%]',
 'U H 3 L1 мгн [%]',
 'U H 4 L1 мгн [%]',
 'U H 5 L1 мгн [%]',
 'U H 6 L1 мгн [%]',
 'U H 7 L1 мгн [%]',
 'U H 8 L1 мгн [%]',
 'U H 9 L1 мгн [%]',
 'U H 10 L1 мгн [%]',
 'U H 11 L1 мгн [%]',
 'U H 12 L1 мгн [%]',
 'U H 13 L1 мгн [%]',
 'U H 14 L1 мгн [%]',
 'U H 15 L1 мгн [%]',
 'U H 16 L1 мгн [%]',
 'U H 17 L1 мгн [%]',
 'U H 18 L1 мгн [%]',
 'U H 19 L1 мгн [%]',
 'U H 20 L1 мгн [%]',
 'U H 21 L1 мгн [%]',
 'U H 22 L1 мгн [%]',
 'U H 23 L1 мгн [%]',
 'U H 24 L1 мгн [%]',
 'U H 25 L1 мгн [%]',
 'U H 26 L1 мгн [%]',
 'U H 27 L1 мгн [%]',
 'U H 28 L1 мгн [%]',
 'U H 29 L1 мгн [%]',
 'U H 30 L1 мгн [%]',
 'U H 31 L1 мгн [%]',
 'U H 32 L1 мгн [%]',
 'U H 33 L1 мгн [%]',
 'U H 34 L1 мгн [%]',
 'U H 35 L1 мгн [%]',
 'U H 36 L1 мгн [%]',
 'U H 37 L1 мгн [%]',
 'U H 38 L1 мгн [%]',
 'U H 39 L1 мгн [%]

Приведем названия столбцов и их порядок к упорядоченному, стандартному виду.

In [8]:
# Cловарь переменных
list_of_variables = {'дата      ' : 'date',
                     'Время (UTC+0)' : 'time',
                     'U L1 мгн [kВ]' : 'U_fks',
                     'I L1 мгн [A]' : 'I_fks',
                     'CF U L1 мгн [---]' : 'CF_U',
                     'CF I L1 мгн [---]' : 'CF_I',
                     'THD U L1 мгн [%]' : 'THD_U',
                     'THD I L1 мгн [%]' : 'THD_I',
                     'U H 1 L1 мгн [%]' : 'U_H_1',
                     'U H 2 L1 мгн [%]' : 'U_H_2',
                     'U H 3 L1 мгн [%]' : 'U_H_3',
                     'U H 4 L1 мгн [%]' : 'U_H_4',
                     'U H 5 L1 мгн [%]' : 'U_H_5',
                     'U H 6 L1 мгн [%]' : 'U_H_6',
                     'U H 7 L1 мгн [%]' : 'U_H_7',
                     'U H 8 L1 мгн [%]' : 'U_H_8',
                     'U H 9 L1 мгн [%]' : 'U_H_9',
                     'U H 10 L1 мгн [%]' : 'U_H_10',
                     'U H 11 L1 мгн [%]' : 'U_H_11',
                     'U H 12 L1 мгн [%]' : 'U_H_12',
                     'U H 13 L1 мгн [%]' : 'U_H_13',
                     'U H 14 L1 мгн [%]' : 'U_H_14',
                     'U H 15 L1 мгн [%]' : 'U_H_15',
                     'U H 16 L1 мгн [%]' : 'U_H_16',
                     'U H 17 L1 мгн [%]' : 'U_H_17',
                     'U H 18 L1 мгн [%]' : 'U_H_18',
                     'U H 19 L1 мгн [%]' : 'U_H_19',
                     'U H 20 L1 мгн [%]' : 'U_H_20',
                     'U H 21 L1 мгн [%]' : 'U_H_21',
                     'U H 22 L1 мгн [%]' : 'U_H_22',
                     'U H 23 L1 мгн [%]' : 'U_H_23',
                     'U H 24 L1 мгн [%]' : 'U_H_24',
                     'U H 25 L1 мгн [%]' : 'U_H_25',
                     'U H 26 L1 мгн [%]' : 'U_H_26',
                     'U H 27 L1 мгн [%]' : 'U_H_27',
                     'U H 28 L1 мгн [%]' : 'U_H_28',
                     'U H 29 L1 мгн [%]' : 'U_H_29',
                     'U H 30 L1 мгн [%]' : 'U_H_30',
                     'U H 31 L1 мгн [%]' : 'U_H_31',
                     'U H 32 L1 мгн [%]' : 'U_H_32',
                     'U H 33 L1 мгн [%]' : 'U_H_33',
                     'U H 34 L1 мгн [%]' : 'U_H_34',
                     'U H 35 L1 мгн [%]' : 'U_H_35',
                     'U H 36 L1 мгн [%]' : 'U_H_36',
                     'U H 37 L1 мгн [%]' : 'U_H_37',
                     'U H 38 L1 мгн [%]' : 'U_H_38',
                     'U H 39 L1 мгн [%]' : 'U_H_39',
                     'U H 40 L1 мгн [%]' : 'U_H_40',
                     'I H 1 L1 мгн [%]' : 'I_H_1',
                     'I H 2 L1 мгн [%]' : 'I_H_2',
                     'I H 3 L1 мгн [%]' : 'I_H_3',
                     'I H 4 L1 мгн [%]' : 'I_H_4',
                     'I H 5 L1 мгн [%]' : 'I_H_5',
                     'I H 6 L1 мгн [%]' : 'I_H_6',
                     'I H 7 L1 мгн [%]' : 'I_H_7',
                     'I H 8 L1 мгн [%]' : 'I_H_8',
                     'I H 9 L1 мгн [%]' : 'I_H_9',
                     'I H 10 L1 мгн [%]' : 'I_H_10',
                     'I H 11 L1 мгн [%]' : 'I_H_11',
                     'I H 12 L1 мгн [%]' : 'I_H_12',
                     'I H 13 L1 мгн [%]' : 'I_H_13',
                     'I H 14 L1 мгн [%]' : 'I_H_14',
                     'I H 15 L1 мгн [%]' : 'I_H_15',
                     'I H 16 L1 мгн [%]' : 'I_H_16',
                     'I H 17 L1 мгн [%]' : 'I_H_17',
                     'I H 18 L1 мгн [%]' : 'I_H_18',
                     'I H 19 L1 мгн [%]' : 'I_H_19',
                     'I H 20 L1 мгн [%]' : 'I_H_20',
                     'I H 21 L1 мгн [%]' : 'I_H_21',
                     'I H 22 L1 мгн [%]' : 'I_H_22',
                     'I H 23 L1 мгн [%]' : 'I_H_23',
                     'I H 24 L1 мгн [%]' : 'I_H_24',
                     'I H 25 L1 мгн [%]' : 'I_H_25',
                     'I H 26 L1 мгн [%]' : 'I_H_26',
                     'I H 27 L1 мгн [%]' : 'I_H_27',
                     'I H 28 L1 мгн [%]' : 'I_H_28',
                     'I H 29 L1 мгн [%]' : 'I_H_29',
                     'I H 30 L1 мгн [%]' : 'I_H_30',
                     'I H 31 L1 мгн [%]' : 'I_H_31',
                     'I H 32 L1 мгн [%]' : 'I_H_32',
                     'I H 33 L1 мгн [%]' : 'I_H_33',
                     'I H 34 L1 мгн [%]' : 'I_H_34',
                     'I H 35 L1 мгн [%]' : 'I_H_35',
                     'I H 36 L1 мгн [%]' : 'I_H_36',
                     'I H 37 L1 мгн [%]' : 'I_H_37',
                     'I H 38 L1 мгн [%]' : 'I_H_38',
                     'I H 39 L1 мгн [%]' : 'I_H_39',
                     'I H 40 L1 мгн [%]' : 'I_H_40',
                     'cos(φ) L1 мгн [---]' : 'cos_fi'}

In [9]:
# Переобозначим переменные 
fks_4 = fks_4.rename(columns = list_of_variables)

In [10]:
# Проверка результата
fks_4.head()

date           time  U_fks   I_fks  CF_U  CF_I  THD_U  THD_I  U_H_1  \
0  08.09.2020  04:22:50.242   14.53  211.92  1.59  1.56   8.86  16.19  99.60   
1  08.09.2020  04:22:51.243   14.51  216.06  1.59  1.58   9.04  15.53  99.58   
2  08.09.2020  04:22:52.243   14.50  217.02  1.59  1.56   9.13  15.46  99.57   
3  08.09.2020  04:22:53.244   14.50  215.73  1.59  1.55   9.11  15.56  99.57   
4  08.09.2020  04:22:54.244   14.49  213.97  1.60  1.57   9.10  15.94  99.58   

   U_H_2  ...  I_H_32  I_H_33  I_H_34  I_H_35  I_H_36  I_H_37  I_H_38  I_H_39  \
0   0.06  ...    0.01    0.24    0.03    0.11    0.03    0.20    0.02    0.23   
1   0.07  ...    0.02    0.30    0.03    0.17    0.03    0.22    0.03    0.26   
2   0.05  ...    0.04    0.31    0.03    0.25    0.03    0.19    0.02    0.27   
3   0.07  ...    0.02    0.37    0.04    0.34    0.03    0.30    0.02    0.29   
4   0.05  ...    0.03    0.37    0.02    0.12    0.03    0.14    0.04    0.22   

   I_H_40  cos_fi  
0    0.02    0.50  
1    0.02    0.50  
2    0.05    0.50  
3    0.03    0.50  
4    0.02    0.49  

[5 rows x 89 columns]

### _Describe data_

---

- data - дата проведения измерения
- time - время проведения измерения 
- U_fks - напряжение фидера 
- I_fks - ток фидера
- CF_U - Коэффициент пиковых значений напряжения
- CF_I - Коэффициент пиковых значений тока
- THD_U - Коэффициент гармонических искажений напряжения
- THD_I - Коэффициент гармонических искажений тока
- U_H_1 : U_H_40 - Амплитуды гармоник напряжения
- I_H_1 : I_H_40 - Амплитуды гармоник тока
- cos_fi - Косинус угла сдвига фаз

## SECTION 2.2 DATA PREPARATION

### _Verify data quality_

---

In [13]:
# Функция коплексной оценки качества данных 
def verify_data_quality (evaluated_dataset):
    # Проверка типов данных
    print("Типы данных:\n{}\n".format(evaluated_dataset.dtypes) + '_' * 100+ '\n')
    # Проверка наличия значений NaN
    print("Пропущенные значения:\n{}\n".format(evaluated_dataset.isnull().sum().sum()) + '_' * 100+ '\n')
    # Количество значений, отличных от NaN
    print("Количество значений, отличных от NaN:\n{}\n".format(evaluated_dataset.count().sum()) + '_' * 100+ '\n')
    # Поиск дублирующих строк 
    print("Дублирующих строк:\n{}\n".format(evaluated_dataset.duplicated().sum()) + '_' * 100+ '\n')
    # Форма нарбора данных
    print("Форма нарбора данных:\n{}\n".format(evaluated_dataset.shape) + '_' * 100+ '\n')
    # Тип индекса набора данных
    print("Тип индекса набора данных:\n{}\n".format(evaluated_dataset.index) + '_' * 100+ '\n')

In [14]:
# Оценка качетва данных
verify_data_quality(fks_4)

Типы данных:
date       object
time       object
U_fks     float64
I_fks     float64
CF_U      float64
           ...   
I_H_37    float64
I_H_38    float64
I_H_39    float64
I_H_40    float64
cos_fi    float64
Length: 89, dtype: object
____________________________________________________________________________________________________

Пропущенные значения:
0
____________________________________________________________________________________________________

Количество значений, отличных от NaN:
88965290
____________________________________________________________________________________________________

Дублирующих строк:
0
____________________________________________________________________________________________________

Форма нарбора данных:
(999610, 89)
____________________________________________________________________________________________________

Тип индекса набора данных:
RangeIndex(start=0, stop=999610, step=1)
__________________________________________________________

### _Data type conversions_

---

Преобразуем типы данных даты и времени для анализа

In [15]:
# Преобразование даты 
fks_4['date'] = pd.to_datetime(fks_4['date'])

In [16]:
# Преобразование времени  
fks_4['time'] = pd.to_datetime(fks_4['time']).dt.time

In [17]:
# Проверка результата
fks_4.head()

date             time  U_fks   I_fks  CF_U  CF_I  THD_U  THD_I  U_H_1  \
0 2020-08-09  04:22:50.242000  14.53  211.92  1.59  1.56   8.86  16.19  99.60   
1 2020-08-09  04:22:51.243000  14.51  216.06  1.59  1.58   9.04  15.53  99.58   
2 2020-08-09  04:22:52.243000  14.50  217.02  1.59  1.56   9.13  15.46  99.57   
3 2020-08-09  04:22:53.244000  14.50  215.73  1.59  1.55   9.11  15.56  99.57   
4 2020-08-09  04:22:54.244000  14.49  213.97  1.60  1.57   9.10  15.94  99.58   

   U_H_2  ...  I_H_32  I_H_33  I_H_34  I_H_35  I_H_36  I_H_37  I_H_38  I_H_39  \
0   0.06  ...    0.01    0.24    0.03    0.11    0.03    0.20    0.02    0.23   
1   0.07  ...    0.02    0.30    0.03    0.17    0.03    0.22    0.03    0.26   
2   0.05  ...    0.04    0.31    0.03    0.25    0.03    0.19    0.02    0.27   
3   0.07  ...    0.02    0.37    0.04    0.34    0.03    0.30    0.02    0.29   
4   0.05  ...    0.03    0.37    0.02    0.12    0.03    0.14    0.04    0.22   

   I_H_40  cos_fi  
0    0.02    0.50  
1    0.02    0.50  
2    0.05    0.50  
3    0.03    0.50  
4    0.02    0.49  

[5 rows x 89 columns]

### _Sorting, reordering, reindexing_

---

In [18]:
# Установим время как индекс 
fks_4 = fks_4.set_index('time')

In [19]:
# Проверка результата
fks_4.head()

date  U_fks   I_fks  CF_U  CF_I  THD_U  THD_I  U_H_1  \
time                                                                         
04:22:50.242000 2020-08-09  14.53  211.92  1.59  1.56   8.86  16.19  99.60   
04:22:51.243000 2020-08-09  14.51  216.06  1.59  1.58   9.04  15.53  99.58   
04:22:52.243000 2020-08-09  14.50  217.02  1.59  1.56   9.13  15.46  99.57   
04:22:53.244000 2020-08-09  14.50  215.73  1.59  1.55   9.11  15.56  99.57   
04:22:54.244000 2020-08-09  14.49  213.97  1.60  1.57   9.10  15.94  99.58   

                 U_H_2  U_H_3  ...  I_H_32  I_H_33  I_H_34  I_H_35  I_H_36  \
time                           ...                                           
04:22:50.242000   0.06   3.35  ...    0.01    0.24    0.03    0.11    0.03   
04:22:51.243000   0.07   3.37  ...    0.02    0.30    0.03    0.17    0.03   
04:22:52.243000   0.05   3.53  ...    0.04    0.31    0.03    0.25    0.03   
04:22:53.244000   0.07   3.72  ...    0.02    0.37    0.04    0.34    0.03   
04:22:54.244000   0.05   3.76  ...    0.03    0.37    0.02    0.12    0.03   

                 I_H_37  I_H_38  I_H_39  I_H_40  cos_fi  
time                                                     
04:22:50.242000    0.20    0.02    0.23    0.02    0.50  
04:22:51.243000    0.22    0.03    0.26    0.02    0.50  
04:22:52.243000    0.19    0.02    0.27    0.05    0.50  
04:22:53.244000    0.30    0.02    0.29    0.03    0.50  
04:22:54.244000    0.14    0.04    0.22    0.02    0.49  

[5 rows x 88 columns]

In [20]:
# Приведение к фазному значению
import math
# Наряжение 
fks_4['U_fks'] = fks_4['U_fks']*math.sqrt(3)

# Ток
# fks_4['I_fks'] = fks_4['I_fks']*math.sqrt(3)

### _Conclusions_

---

В текущем разделе проекта выполнен сбор и первичный анализ результатов измерениий. 

- Исходный набор данных имеет форму: (999610, 89). 
- Заголовки переменных имеют проблы и лишние символы и обозначения. В связи с чем обозначения переменных приведенно к принятому в анализе виду. 
- Весь набор имеет два типа данных: 
    - object
    - float64
- Типом object представлены перменные date и time. Данные перемены приведены к соответвующим форматам даты и времени 
- Переменная time установленна в качестве индекса.
- Оценка качества данных показала, что в наборе нет пропусков и дублирующих значений.

In [32]:
# Сохранение результатов предварительной обработки для дальнейшей работы 
fks_4.to_csv('fks_4_prepared.csv', sep=',', index=True, encoding='utf-8')